util/chunk/codec.go

```diff
@@ -189,3 +189,139 @@ func init() {
+//    2.2 Append srcCol.offsets to destCol.offsets when the elements is of var-length type. And further adjust the
+//        offsets according to descCol.offsets[destCol.length]-srcCol.offsets[0].
+//    2.3 Append srcCol.nullBitMap to destCol.nullBitMap.
+// 3. Go to step 1 when the input byte slice is consumed.
+type Decoder struct {
+	intermChk    *Chunk
+	colTypes     []*types.FieldType
+	remainedRows int
+}
+
+// NewDecoder creates a new Decoder object for decode a Chunk.
+func NewDecoder(chk *Chunk, colTypes []*types.FieldType) *Decoder {
+	return &Decoder{intermChk: chk, colTypes: colTypes, remainedRows: 0}
+}
+
+// Decode decodes multiple rows of Decoder.intermChk and stores the result in chk.
+func (c *Decoder) Decode(chk *Chunk) {
+	requiredRows := chk.RequiredRows() - chk.NumRows()
+	// Set the requiredRows to a multiple of 8.
+	requiredRows = (requiredRows + 7) >> 3 << 3
+	if requiredRows > c.remainedRows {
+		requiredRows = c.remainedRows
+	}
+	for i := 0; i < len(c.colTypes); i++ {
+		c.decodeColumn(chk, i, requiredRows)
+	}
+	c.remainedRows -= requiredRows
+}
+
+// Reset decodes data and store the result in Decoder.intermChk. This decode phase uses pointer operations with less
+// CPU and memory costs.
+func (c *Decoder) Reset(data []byte) {
+	codec := NewCodec(c.colTypes)
+	codec.DecodeToChunk(data, c.intermChk)
+	c.remainedRows = c.intermChk.NumRows()
+}
+
+// IsFinished indicates whether Decoder.intermChk has been dried up.
+func (c *Decoder) IsFinished() bool {
+	return c.remainedRows == 0
+}
+
+func (c *Decoder) decodeColumn(chk *Chunk, ordinal int, requiredRows int) {
+	elemLen := getFixedLen(c.colTypes[ordinal])
+	numDataBytes := int64(elemLen * requiredRows)
+	srcCol := c.intermChk.columns[ordinal]
+	destCol := chk.columns[ordinal]
+
+	if elemLen == varElemLen {
+		// For var-length types, we need to adjust the offsets after appending to descCol.
+		numDataBytes = srcCol.offsets[requiredRows] - srcCol.offsets[0]
+		deltaOffset := destCol.offsets[destCol.length] - srcCol.offsets[0]
+		destCol.offsets = append(destCol.offsets, srcCol.offsets[1:requiredRows+1]...)
+		for i := destCol.length + 1; i <= destCol.length+requiredRows; i++ {
+			destCol.offsets[i] = destCol.offsets[i] + deltaOffset
+		}
+		srcCol.offsets = srcCol.offsets[requiredRows:]
+	}
+
+	numNullBitmapBytes := (requiredRows + 7) >> 3
+	if destCol.length%8 == 0 {
+		destCol.nullBitmap = append(destCol.nullBitmap, srcCol.nullBitmap[:numNullBitmapBytes]...)
+	} else {
+		destCol.appendMultiSameNullBitmap(false, requiredRows)
+		bitMapLen := len(destCol.nullBitmap)
+		// bitOffset indicates the number of valid bits in destCol.nullBitmap's last byte.
+		bitOffset := destCol.length % 8
+		startIdx := (destCol.length - 1) >> 3
+		for i := 0; i < numNullBitmapBytes; i++ {
+			destCol.nullBitmap[startIdx+i] |= srcCol.nullBitmap[i] << bitOffset
+			// the total number of elements maybe is small than bitMapLen * 8, and last some bits in srcCol.nullBitmap are useless.
```

Comment:
The high order 8-bitOffset bits in \`srcCol.nullBitmap[i]\` should be appended to the low order of the next slot.